types/types.go

```diff
@@ -265,7 +265,7 @@ type ImageDestination interface {
 	// WARNING: The contents of stream are being verified on the fly.  Until stream.Read() returns io.EOF, the contents of the data SHOULD NOT be available
 	// to any other readers for download using the supplied digest.
 	// If stream.Read() at any time, ESPECIALLY at end of input, returns an error, PutBlob MUST 1) fail, and 2) delete any data stored so far.
-	PutBlob(ctx context.Context, stream io.Reader, inputInfo BlobInfo, cache BlobInfoCache, isConfig bool) (BlobInfo, error)
+	PutBlob(ctx context.Context, stream io.Reader, inputInfo BlobInfo, layerIndexInImage int, cache BlobInfoCache, isConfig bool) (BlobInfo, error)
```

Comment:
(Warning, vague:)

This is an ABI break, i.e. we should use a new major version number in the next release.

1. Do we just do it, without second thought? https://github.com/containers/image/pull/609 seems to suggest so.
2. Do we somehow say that (parts of?) the transport interfaces are going to break even within a major version, and only things like `copy.Image` and `alltransports.ParseReference` (`docker.NewReference`?) are stable for now? How would the callers know which one is which? (Also, I do know that at least for reading there are third-party callers of the `ImageSource` interface.)
3. Do we want to now spend extra time thinking about `PutBlob` (or, worse, ~everything here) to make future ABI breaks less likely? Like maybe moving (`layerIndexInImage`, `cache`, `isConfig`) into an extensible `PutBlobAdditionalData` structure that will allow us to add (but not remove!) members in the future without an ABI break?

I don’t really know…
- The lazy me would prefer 1.
- Something like 2. would be much easier to sustain than a full ABI but awkward to stomach for some users (in the extreme, we could make the transport interfaces `internal` and break those users completely, to have a clean ABI — clearly not worth it but to an extent attractive)
- 3\. at a first glance seems like a fair amount of effort that is still likely enough to fail anyway, as the surrounding assumptions change (e.g. suddenly transports having to cope with running in a user namespace), or make the interfaces notably less readable and more difficult to maintain, considering how quickly have these interfaces (and `PutBlob`/`TryReusingBlob` in particular) been changing.

(It’s been tempting to contemplate a variant of 2., splitting c/image into an ABI-stable Go module and an ABI-unstable v0 module, or something like that, to be very transparent to users about what they can/can’t expect with a particular symbol; but the ABI-stable calls to the unstable v0 would have to be tightly coupled, and I can’t see how to do this with Go modules so that an import of the ABI-stable module automatically brings in the corresponding unstable v0 version, _but_ so that we don’t have to maintain two independent c/image sub-repos (with independent tests, PRs, tags) or something similar. It didn’t seem viable IIRC.)

Reply:
(To be clear, none of this is in any way at all a reason to reject this PR; it’s just that we do need to actually set a precedent for what #609 means.)

Reply:
> This is an ABI break, i.e. we should use a new major version number in the next release.

Starting with v2, we have to consider the following go-mod detail:

_If the module is version v2 or higher, the major version of the module must be included as a /vN at the end of the module paths used in go.mod files (e.g., module github.com/my/mod/v2, require github.com/my/mod/v2 v2.0.0) and in the package import path (e.g., import "github.com/my/mod/v2/mypkg")._

_If the module is version v0 or v1, do not include the major version in either the module path or the import path._ [[go wiki](https://github.com/golang/go/wiki/Modules#semantic-import-versioning)]

This will affect us as soon as we migrate to `go mod`. Users of containers/image who already migrated to `go mod` need to mark it as `+incompatible` and will not be interrupted any further.

> (To be clear, none of this is in any way at all a reason to reject this PR; it’s just that we do need to actually set a precedent for what #609 means.)

No worries, I really appreciate that you kick off a discussion around this topic.

In the context of this PR, I would go on and merge it so we can progress on the blob locking. Coming up with a plan and executing it sounds time-consuming and will likely entail some changes to the API (and the visibility of packages) in any case, so I think it's ~okay to perform the change here. I have opened PRs for Skopeo and Buildah and will open some for Podman and CRI-O as well to have a minimal interruption window for others who might run into build errors when using the latest c/image.

1. Personally, I don't find it to be sustainable over time. Breaking APIs is not only a bad experience for users but it can also be very time consuming and tedious to propagate those changes across our projects; sometimes even blocking other work in case of build errors etc. Having a stable APIs across the stack + `go mod` sounds so sweet (and a bit utopian) :)

2. I think that moving packages into `./internal` is a core part of migrating to `go mod` (and doing semver for go in general). We need to peform (breaking) changes at some point but also need to structure the code into packages, and we can do both with `./internal` packages. Analogous to the kernel's stable syscall interface and the unstable internal ones.

3. c/image has been quite stable. However, this PR suggests that the APIs are hard to extend. Your proposal 3) is aiming at this problem. However, even if we had a dedicated structure to encapsulate the arguments, we still had to do a major version bump as the semantics have changed (at least for the storage destination). It is still a major gain over creating build errors.

My perference is a combination of 2) and 3) as I think both are somehow mandatory for how packages in go work conceptually and how their visibility impacts semver (and go modules). But I did not think this through entirely, so take my reply with a grain of salt.

I suggest asking to give this topic a high priority so we can allocate the required time in the upcoming sprints. Parts of the tasks seem to have a rather conceptual nature of how to do things (e.g., `./internal`) but there is also the massive technical part of analyzing which APIs need to be external and how we can re-design them to be easier to maintain and introduce less build breakage (as in the case of this PR).

I find this a *very* important discussion and once migrated to `go mod` it will affect other projects as well. Hence, pinging the Jedi council for future sprint planning and a general FYI (Cc @mrunalp @rhatdan).

Reply:
@mtrmac what is the issue with doing semver and incrementing the major number?

In the end, it is way easier to understand the nature of updating a dependency that is doing semver. The benefit we gained doing semver is on the user side as `go mod` can update the dependencies automatically while having a non-semver version scheme forces users to do the same tedious work as with `vndr`.

Reply:
> (Unless there is a clear settled answer, maybe we should break this out into a separate issue?)

We can make this a PS in tomorrow's meeting and/or set up a dedicated one to discuss it. To a certain degree, the discussion conceptually affects the entire github.com/containers/* repos and having clear answers to the questions would be beneficial for a common and consistent release engineering.

Reply:
> @mtrmac what is the issue with doing semver and incrementing the major number?

I was reading the following as you not wanting to do it:
> In the context of this PR, I would go on and merge it so we can progress on the blob locking. Coming up with a plan and executing it sounds time-consuming and will likely entail some changes to the API (and the visibility of packages) in any case, so I think it's ~okay to perform the change here. 

Was that incorrect?

---

BTW option 3., just moving the data into an extensible `struct` would not really help, if `storageImageDestination` now requires the layer index to be present, we would effectively be breaking the API even without breaking compilation.

Reply:
> Was that incorrect?

I am sorry, I did not realize that my statement was vague.

What I meant by  _"I would go on and merge it so we can progress on the blob locking"_ was that I prefer to merge this PR although it breaks the API and will entail a major version bump. In other words, I am okay with a major version bump for this particular case (as I don't see any other way to achieve the blob locking).

Reply:
Ah, OK. Let’s tentatively plan to just bump the major version, then.